qemu/tests/block_copy.py

```diff
@@ -456,3 +449,50 @@ def reopen(self, reopen_image):
         self.params["image_name_%s" % self.tag] = reopen_image
         self.vm.create(params=self.params)
         self.vm.verify_alive()
+
+    def hot_unplug(self):
+        """
+        Host unplug the source image, check if device deleted and block job cancelled
+        or completed, both results is acceptable per different qemu version.
+        """
+        job_cancelled_events = ["BLOCK_JOB_CANCELLED", "BLOCK_JOB_COMPLETED"]
+        device_delete_event = ["DEVICE_DELETED"]
+        for event in job_cancelled_events + device_delete_event:
+            self.vm.monitor.clear_event(event)
+
+        def is_unplugged():
+            event_list = self.vm.monitor.get_events()
+            logging.debug("event_list: %s" % event_list)
+            device_deleted = job_cancelled = False
+            for event_str in event_list:
+                event = event_str.get("event")
+                if event in device_delete_event:
+                    device_deleted = True
+                if event in job_cancelled_events:
+                    job_cancelled = True
+            return device_deleted and job_cancelled
+
+        qdev = self.vm.devices
+        device = qdev.get_by_params({"drive": self.device})
+        if not device:
+            self.test.fail("Device does not exist.")
+        logging.info("Hot unplug device %s" % self.device)
+        qdev.simple_unplug(device[0], self.vm.monitor)
+        timeout = self.params.get("cancel_timeout", 10)
+        unplugged = utils_misc.wait_for(is_unplugged, timeout=timeout)
+        if not unplugged:
+            self.test.fail("Unplug timeout in %ss" % timeout)
```

Comment:
Maybe, the result is not clear enough.

Reply:
@yanan-fu It is hard to separate the check of all events, since sometimes they prompt very quick, if catch them one by one, we will miss them sometimes. Luckily we can check all qmp prompts after the test finished if it failed. So allow me keep it like this, thanks.